### PR TITLE
feat(contract): points for membership fee

### PR DESCRIPTION
**Pull request by shuhuiluo**
Introduce the `RiverPoints` contract implementing the ERC20 standard with metadata using associated storage, balance, and allowance management libraries. These changes establish the foundational structure for handling `RiverPoints` tokens securely and efficiently.

Replaced `river` with `spaceFactory` in `RiverPoints` to support space-based minting. Added new functionality in `MembershipJoin` to charge join fees, handle partner and referral fees, and credit points accordingly. Enhanced security and modularity by introducing the `onlySpace` modifier and custom error handling.

### CODE REVIEW

**Comment by giuseppecrj** [comment]
Why are we using a custom ERC20 implementation instead of a standard?

File: contracts/src/spaces/facets/membership/join/MembershipJoin.sol [hunk 1]
```diff
@@ -260,32 +263,71 @@ abstract contract MembershipJoin is
...
+    RiverPoints pointsToken;
+    // Equivalent to `pointsToken.mint(receiver, points);`
+    bytes4 selector = RiverPoints.mint.selector;
+    assembly ("memory-safe") {
+      mstore(0, selector)
+      mstore(0x04, receiver)
+      mstore(0x24, points)
+      pop(call(gas(), pointsToken, 0, 0, 0x44, 0, 0))
+      mstore(0x24, 0)
+    }
```
[the remark this answers is not in the document]
Comment (shuhuiluo) [reply contracts/src/spaces/facets/membership/join/MembershipJoin.sol:310]: It will fail existing tests if checking for success.

**Comment by shuhuiluo** [comment]
> Why are we using a custom ERC20 implementation instead of a standard?

Because there's no `ERC20Facet` and `ERC20Upgradeable` is an overkill.

File: contracts/src/spaces/facets/membership/join/MembershipJoin.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (giuseppecrj) [reply contracts/src/spaces/facets/membership/join/MembershipJoin.sol:310]: update tests then.

File: contracts/src/spaces/facets/membership/join/MembershipJoin.sol [hunk 2]
```diff
@@ -221,8 +222,8 @@ abstract contract MembershipJoin is
   /// @notice Processes the charge for joining a space without referral
   /// @param transactionId The unique identifier for this join transaction
   function _chargeForJoinSpace(bytes32 transactionId) internal {
-    uint256 membershipPrice = _getCapturedValue(transactionId);
-    if (membershipPrice == 0) revert Membership__InsufficientPayment();
+    uint256 payment = _getCapturedValue(transactionId);
```
Comment (giuseppecrj) [inline contracts/src/spaces/facets/membership/join/MembershipJoin.sol:225]: Move renaming to separate PR

File: contracts/src/spaces/facets/membership/join/MembershipJoin.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (shuhuiluo) [reply contracts/src/spaces/facets/membership/join/MembershipJoin.sol:310]: I don't think failure to mint points should block membership minting.

**Comment by shuhuiluo** [comment]
> Why are we using a custom ERC20 implementation instead of a standard?

@giuseppecrj Making an `ERC20Facet` is also an overkill. I don't want to inherit `ERC20Facet` and override every function.

`ERC20Lib` is a rewrite of OpenZeppelin `ERC20Upgradeable`, courtesy of [solitude](https://github.com/shuhuiluo/solitude).

File: contracts/src/spaces/facets/membership/join/MembershipJoin.sol (hunk 2 above)
Comment (shuhuiluo) [reply contracts/src/spaces/facets/membership/join/MembershipJoin.sol:225]: Yes sir.

**Comment by giuseppecrj** [comment]
> `20Facet` and override every function.

Create ERC20Base for this, same as ERC721ABase.sol

**Comment by shuhuiluo** [comment]
> > `20Facet` and override every function.
> 
> Create ERC20Base for this, same as ERC721ABase.sol

`ERC20Base` will just be a wrapper of `ERC20Lib`.

File: contracts/src/spaces/facets/membership/join/MembershipJoin.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (giuseppecrj) [reply contracts/src/spaces/facets/membership/join/MembershipJoin.sol:309]: I disagree, please update appropiately.

File: contracts/src/primitive/AllowanceMap.sol [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+using AllowanceLib for AllowanceMap global;
```
Comment (giuseppecrj) [inline contracts/src/primitive/AllowanceMap.sol:4]: can we avoid using global and instead declare it in the RiverPoints contract directly? If we have different versions of AllowanceLib in different facets, can this cause conflicts?

It makes the dependencies more explicit, which is crucial for diamond proxy patterns.

**Review by giuseppecrj — CHANGES_REQUESTED** [review]

**Review by pfives — APPROVED** [review]

**Pull request merged**